### PR TITLE
Fixup intpromote by default

### PR DESCRIPTION
**Pull request by dkorpel**
Address the review of https://github.com/dlang/dmd/pull/13348
@ibuclaw @schveiguy

### CODE REVIEW

File: changelog/fix16997.dd [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
-~byte as typeof(byte)(~int(byte)) to preserve the old behavior or use -revert=intpromote to revert the changes
+---
+void main()
+{
+    // Note: byte.max = 127, byte.min = -128
+    byte b = -128;
+    int x = -b; // new behavior: x = 128
+    int y = cast(byte)(-b); // old behavior: y = -128
+}
+---
```
Comment (schveiguy) [inline changelog/fix16997.dd:17]: It should also note things that worked with the old behavior and now require a cast. Like
```d
byte b2 = cast(byte)-b; // cast now required
```

Comment (schveiguy) [reply changelog/fix16997.dd:18]: OK, it's good enough.

File: src/dmd/frontend.h [hunk 2]
```diff
@@ -707,7 +708,7 @@ struct Param final
...
...level = 0u, Array<const char* >* debugids = nullptr, uint32_t versionlevel = 0u, Array<const char* >* versionids = nullptr, _d_dynamicArray< const char > defaultlibname = {}, _d_dynamicArray< const char > debuglibname = {}, _d_dynamicArray< const char > mscrtlib = {}, _d_dynamicArray< const char > moduleDepsFile = {}, OutBuffer* moduleDeps = nullptr, bool emitMakeDeps = false, _d_dynamicArray< const char > makeDepsFile = {}, Array<const char* > makeDeps = Array<const char* >(0LLU, {}, arrayliteral), MessageStyle messageStyle = (MessageStyle)0u, bool run = false, Array<const char* > runargs = Array<const char* >(0LLU, {}, arrayliteral), Array<const char* > objfiles = Array<const char* >(0LLU, {}, arrayliteral), Array<const char* > linkswitches = Array<const char* >(0LLU, {}, arrayliteral), Array<bool > linkswitchIsForCC = Array<bool >(0LLU, {}, arrayliteral), Array<const char* > libfiles = Array<const char* >(0LLU, {}, arrayliteral), Array<const char* > dllfiles = Array<const char* >(0LLU, {}, arrayliteral), _d_dynamicArray< const char > deffile = {}, _d_dynamicArray< const char > resfile = {}, _d_dynamicArray< const char > exefile = {}, _d_dynamicArray< const char > mapfile = {}) :
```
Comment (12345swordy) [inline src/dmd/frontend.h:711]: Sidenote: We really need a better way of generating this as it is way too wide for any advantage computer screen out there.

File: test/compilable/previewhelp.d [hunk 3]
```diff
@@ -11,7 +11,6 @@ Upcoming language changes listed by -preview=name:
   =dip1021          implement https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1021.md (Mutable function arguments)
   =fieldwise        use fieldwise comparisons for struct equality
   =fixAliasThis     when a symbol is resolved, check alias this scope before going to upper scopes
-  =intpromote       fix integral promotions for unary + - ~ operators
```
Comment (12345swordy) [inline test/compilable/previewhelp.d:14]: Instead of removing it why not mark it as deprecated?

**Review by 12345swordy — APPROVED** [review]
Other then a couple of nitpicks, this looks good to me.

File: test/compilable/previewhelp.d (hunk 3 above)
Comment (dkorpel) [reply test/compilable/previewhelp.d:14]: It is marked as deprecated in cli.d, but also as undocumented:
```D
Feature("intpromote", "fix16997",
            "fix integral promotions for unary + - ~ operators"),
            "fix integral promotions for unary + - ~ operators", false, true),
```
Which is what Iain suggested:

> You're not removing it, just explicitly undocumenting it, and optionally warning about its use.

No use in listing a useless switch.

File: src/dmd/frontend.h (hunk 2 above)
Comment (ibuclaw) [reply src/dmd/frontend.h:711]: The autogenerated header is mostly for ignoring (other than ensuring that any changes here are reflected properly in the hand-maintained headers).

File: test/compilable/previewhelp.d (hunk 3 above)
Comment (ibuclaw) [reply test/compilable/previewhelp.d:14]: Right, there are two use cases:
1. `false, false` - this switch is redundant, so no need to document it anymore, however we accept that there may be many uses of in the wild and don't want to immediately break downstream builds just yet.
2. `false, true` - same as above, but start warning downstream projects that it is time to move off this switch.

**Review by ljmf00 — CHANGES_REQUESTED** [review]
Can we add tests for deprecation message for `-transition=intpromote` and the changed line about `-transition=16997`?

**Comment by dkorpel** [comment]
> Can we add tests for deprecation message for `-transition=intpromote` and the changed line about `-transition=16997`?

I didn't want to slow down the test suite by adding a test case for a trivial obsolete undocumented switch, but I can add one if you want.

**Review by ljmf00 — APPROVED** [review]
> > Can we add tests for deprecation message for `-transition=intpromote` and the changed line about `-transition=16997`?
> 
> I didn't want to slow down the test suite by adding a test case for a trivial obsolete undocumented switch, but I can add one if you want.

As far as I know, the transition flag is not an obsolete or undocumented switch. IMO, if we can easily test something, we should do it in favour of increasing/maintaining the coverage percentage, although I don't really know the rule of thumb in DMD, specifically about trivial testing. Anyway, I don't really want to make this a blocker, just take it as a suggestion :)

**Comment by dkorpel** [comment]
> As far as I know, the transition flag is not an obsolete or undocumented switch. 

The code says:
```D
                // Legacy -transition flags
                // Before DMD 2.085, DMD `-transition` was used for all language flags
                // These are kept for backwards compatibility, but no longer documented
```
intpromote was only documented as a `-preview=` flag
```
dmd -transition=h
Language transitions listed by -transition=name:
  =all              Enables all available language transitions
  =field            list all non-mutable fields which occupy an object instance
  =tls              list all variables going into thread local storage
  =vmarkdown        list instances of Markdown replacements in Ddoc
```

> IMO, if we can easily test something, we should do it in favour of increasing/maintaining the coverage percentage

I agree with the importance of test coverage, I was just conflicted in this case because the dmd test suite is big and slow and we keep piling onto it so I thought we could let this case slide :3

**Comment by ljmf00** [comment]
> > As far as I know, the transition flag is not an obsolete or undocumented switch.
> 
> The code says:
> 
> ```d
>                 // Legacy -transition flags
>                 // Before DMD 2.085, DMD `-transition` was used for all language flags
>                 // These are kept for backwards compatibility, but no longer documented
> ```
> 
> intpromote was only documented as a `-preview=` flag
> 
> ```
> dmd -transition=h
> Language transitions listed by -transition=name:
>   =all              Enables all available language transitions
>   =field            list all non-mutable fields which occupy an object instance
>   =tls              list all variables going into thread local storage
>   =vmarkdown        list instances of Markdown replacements in Ddoc
> ```
> 
> > IMO, if we can easily test something, we should do it in favour of increasing/maintaining the coverage percentage
> 
> I agree with the importance of test coverage, I was just conflicted in this case because the dmd test suite is big and slow and we keep piling onto it so I thought we could let this case slide :3

Ah ok, thanks for clarifying :) Yeah, I understand the tradeoff.

**Comment by ibuclaw** [comment]
> Ah ok, thanks for clarifying :) Yeah, I understand the tradeoff.

Neither here nor there about it really.

One the one hand, why add new tests for private (explicitly undocumented) code, in particular when it's been that way for years.

On the other hand, the testsuite is a junkyard of tests that no reasonable human would try out anyway, and we have them just because you can, not because they have a valid purpose.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**